### PR TITLE
Replace State tokio::sync::Mutex by std::sync::RwLock

### PR DESCRIPTION
**Pull request by rbran**
Using RwLock allow multiple tasks to read the value contained in parallel. In the future, this could be improved even further by using internal mutability on the internal structure. Eg using a read lock on state and a write lock on audio, this way is possible to other task access phase_watcher in parallel.

From https://doc.rust-lang.org/std/sync/struct.RwLock.html

> In comparison, a Mutex does not distinguish between readers or writers that acquire the lock, therefore blocking any threads waiting for the lock to become available. An RwLock will allow any number of readers to acquire the lock as long as a writer is not holding the lock.

The tokio::sync::Mutex is also more expensive than a regular std::sync::Mutex: https://tokio.rs/tokio/tutorial/shared-state

> That said, an asynchronous mutex is more expensive than an ordinary mutex, and it is typically better to use one of the two other approaches.

I just add `unwrap` in RwLock because the code was filled with them, I suppose this should be fixed in the future.

### CODE REVIEW

File: mumd/src/network/udp.rs [hunk 1]
```diff
@@ -9,7 +9,7 @@ use mumble_protocol::crypt::ClientCryptState;
 use mumble_protocol::ping::{PingPacket, PongPacket};
 use mumble_protocol::voice::VoicePacket;
 use mumble_protocol::Serverbound;
-use std::collections::HashMap;
```
Comment (default-username-852) [inline mumd/src/network/udp.rs:12]: The import of `sync::RwLock` should be grouped below with the import of `std::sync::Arc` and not with the import of `std::collections::HashMap`.

File: mumd/src/network/tcp.rs [hunk 2]
```diff
@@ -103,7 +103,7 @@ pub async fn handle(
         .await?;
 
         // Handshake (omitting `Version` message for brevity)
-        let state_lock = state.lock().await;
+        let state_lock = state.read().unwrap();
         let username = state_lock.username().unwrap().to_string();
         let password = state_lock.password().map(|x| x.to_string());
         authenticate(&mut sink, username, password).await?;
```
Comment (default-username-852) [inline mumd/src/network/tcp.rs:109]: There is a potential risk of a deadlock here. If the lock is acquired and the call to authenticate makes the runtime switch context to another task that also wants to acquire the lock, then the process could be blocked and not be able to resolve. To fix this, you simply need to make sure to drop the lock before awaiting it, and reacquiring it after awaiting.

File: mumd/src/network/tcp.rs [hunk 3]
```diff
@@ -10,7 +10,7 @@ use mumble_protocol::control::{msgs, ClientControlCodec, ControlCodec, ControlPa
 use mumble_protocol::crypt::ClientCryptState;
 use mumble_protocol::voice::VoicePacket;
 use mumble_protocol::{Clientbound, Serverbound};
-use std::collections::HashMap;
```
Comment (default-username-852) [inline mumd/src/network/tcp.rs:13]: This import should also be grouped with the other `std::sync` import.

**Review by default-username-852 — CHANGES_REQUESTED** [review]

**Comment by default-username-852** [comment]
The unwraps should be safe to do. The only time that the calls to `read` and `write` return an `Err` is when the lock has been poisoned, ie when another thread has panic-ed while holding a lock. Since we shouldn't panic, it should be safe.

File: mumd/src/network/tcp.rs (hunk 2 above)
Comment (rbran) [reply mumd/src/network/tcp.rs:109]: Do you think is better to have a drop call?
```
let state_lock = state.read().unwrap();
let username = state_lock.username().unwrap().to_string();
let password = state_lock.password().map(|x| x.to_string());
drop(state_lock);
```
Or a change context?
```
let (username, password) = {
    let state_lock = state.read().unwrap();
    (state_lock.username().unwrap().to_string(),
        state_lock.password().map(|x| x.to_string()))
};
```
First one seems more likely to induce a mistake that cause a deadlock.

Comment (default-username-852) [reply mumd/src/network/tcp.rs:109]: I think I prefer the second alternative, but both seem fine to me.

**Review by default-username-852 — APPROVED** [review]
Great job!

**Review by sornas — APPROVED** [review]

**Pull request merged**